Mathlib/FieldTheory/NormalClosure.lean

```diff
@@ -71,12 +82,75 @@ instance is_finiteDimensional [FiniteDimensional F K] :
+
+instance : IsScalarTower F K (normalClosure F K L) := by
+  apply of_algebraMap_eq'
+  ext x
+  exact algebraMap_apply F K L x
+
+instance : IsScalarTower K (normalClosure F K L) L :=
+  of_algebraMap_eq' rfl
+
+lemma restrictScalars_eq :
+    (toAlgHom K (normalClosure F K L) L).fieldRange.restrictScalars F = normalClosure F K L :=
+  SetLike.ext' Subtype.range_val
 
 end normalClosure
+
+namespace IntermediateField
+
+variable {F L}
+variable (K : IntermediateField F L)
+
+lemma le_normalClosure : K ≤ normalClosure F K L :=
+K.fieldRange_val.symm.trans_le K.val.fieldRange_le_normalClosure
+
+lemma normalClosure_of_normal [Normal F K] : normalClosure F K L = K :=
+by simp only [normalClosure_def, AlgHom.fieldRange_of_normal, iSup_const]
+
+variable [Normal F L]
+
+lemma normalClosure_def' : normalClosure F K L = ⨆ f : L →ₐ[F] L, K.map f := by
+  refine' (normalClosure_def F K L).trans (le_antisymm (iSup_le (fun f ↦ _)) (iSup_le (fun f ↦ _)))
+  · exact le_iSup_of_le (f.liftNormal L) (fun b ⟨a, h⟩ ↦ ⟨a, a.2, h ▸ f.liftNormal_commutes L a⟩)
+  · exact le_iSup_of_le (f.comp K.val) (fun b ⟨a, h⟩ ↦ ⟨⟨a, h.1⟩, h.2⟩)
+
+lemma normalClosure_def'' : normalClosure F K L = ⨆ f : L ≃ₐ[F] L, K.map f := by
+  refine' (normalClosure_def' K).trans (le_antisymm (iSup_le (fun f ↦ _)) (iSup_le (fun f ↦ _)))
+  · exact le_iSup_of_le (f.restrictNormal' L)
+      (fun b ⟨a, h⟩ ↦ ⟨a, h.1, h.2 ▸ f.restrictNormal_commutes L a⟩)
+  · exact le_iSup_of_le f le_rfl
+
+variable {K : IntermediateField F L}
+
+lemma normal_iff_normalClosure_eq : Normal F K ↔ normalClosure F K L = K :=
+⟨@normalClosure_of_normal F L _ _ _ K, fun h ↦ h ▸ normalClosure.normal F K L⟩
+
+lemma normal_iff_normalClosure_le : Normal F K ↔ normalClosure F K L ≤ K :=
+normal_iff_normalClosure_eq.trans (le_normalClosure K).le_iff_eq.symm
+
+lemma normal_iff_forall_fieldRange_le : Normal F K ↔ ∀ σ : K →ₐ[F] L, σ.fieldRange ≤ K :=
+by rw [normal_iff_normalClosure_le, normalClosure_def, iSup_le_iff]
+
+lemma normal_iff_forall_map_le : Normal F K ↔ ∀ σ : L →ₐ[F] L, K.map σ ≤ K :=
+by rw [normal_iff_normalClosure_le, normalClosure_def', iSup_le_iff]
+
+lemma normal_iff_forall_map_le' : Normal F K ↔ ∀ σ : L ≃ₐ[F] L, K.map ↑σ ≤ K :=
+by rw [normal_iff_normalClosure_le, normalClosure_def'', iSup_le_iff]
+
+lemma normal_iff_forall_fieldRange_eq : Normal F K ↔ ∀ σ : K →ₐ[F] L, σ.fieldRange = K :=
+⟨@AlgHom.fieldRange_of_normal F L _ _ _ K, normal_iff_forall_fieldRange_le.2 ∘ fun h σ ↦ (h σ).le⟩
+
+lemma normal_iff_forall_map_eq : Normal F K ↔ ∀ σ : L →ₐ[F] L, K.map σ = K := by
+  refine' ⟨fun h σ ↦ K.fieldRange_val ▸ _, fun h ↦ normal_iff_forall_map_le.2 (fun σ ↦ (h σ).le)⟩
+  rw [AlgHom.map_fieldRange, normal_iff_forall_fieldRange_eq.1 h, fieldRange_val]
+
+lemma normal_iff_forall_map_eq' : Normal F K ↔ ∀ σ : L ≃ₐ[F] L, K.map ↑σ = K :=
+⟨fun h σ ↦ normal_iff_forall_map_eq.1 h σ, fun h ↦ normal_iff_forall_map_le'.2 (fun σ ↦ (h σ).le)⟩
```

Comment:
Would it make sense to make these into a TFAE?

Reply:
It's a little longer, but looks cleaner: https://github.com/leanprover-community/mathlib4/pull/6163/commits/0471830b4d449b984bc84a01c8b42d1875ff7c62